pkg/config/config.go

```diff
@@ -2,492 +2,233 @@ package config
-
-	// IP address of the node, passed to the kubelet.
-	// If not specified, kubelet will use the node's default IP address.
-	NodeIP string `json:"nodeIP"`
-}
-
-type Debugging struct {
-	// Valid values are: "Normal", "Debug", "Trace", "TraceAll".
-	// Defaults to "Normal".
-	LogLevel string `json:"logLevel"`
-}
-
-func GetConfigFile() string {
-	return configFile
-}
-
-func GetDataDir() string {
-	return dataDir
-}
-
-func GetManifestsDir() []string {
-	return manifestsDir
-}
-
-// KubeConfigID identifies the different kubeconfigs managed in the DataDir
-type KubeConfigID string
-
-const (
-	KubeAdmin               KubeConfigID = "kubeadmin"
-	KubeControllerManager   KubeConfigID = "kube-controller-manager"
-	KubeScheduler           KubeConfigID = "kube-scheduler"
-	Kubelet                 KubeConfigID = "kubelet"
-	ClusterPolicyController KubeConfigID = "cluster-policy-controller"
-	RouteControllerManager  KubeConfigID = "route-controller-manager"
-)
-
-// KubeConfigPath returns the path to the specified kubeconfig file.
-func (cfg *MicroshiftConfig) KubeConfigPath(id KubeConfigID) string {
-	return filepath.Join(dataDir, "resources", string(id), "kubeconfig")
-}
+	Debugging Debugging  `json:"debugging"`
 
-func (cfg *MicroshiftConfig) KubeConfigAdminPath(id string) string {
-	return filepath.Join(dataDir, "resources", string(KubeAdmin), id, "kubeconfig")
+	// Internal-only fields
+	Ingress      IngressConfig `json:"-"`
+	userSettings *Config       `json:"-"` // the values read from the config file
 }
 
-var allHostnames []string
-
-func getAllHostnames() ([]string, error) {
-	if len(allHostnames) != 0 {
-		return allHostnames, nil
+// NewDefault creates a new Config struct populated with the
+// default values and with any computed values updated based on those
+// defaults.
+func NewDefault() *Config {
+	c := &Config{}
+	if err := c.fillDefaults(); err != nil {
+		klog.Fatalf("Failed to initialize config: %v", err)
 	}
-	cmd := exec.Command("/bin/hostname", "-A")
-	var out bytes.Buffer
-	cmd.Stdout = &out
-	err := cmd.Run()
-	if err != nil {
-		return nil, fmt.Errorf("Error when executing 'hostname -A': %v", err)
+	if err := c.updateComputedValues(); err != nil {
+		klog.Fatalf("Failed to initialize config: %v", err)
 	}
-	outString := out.String()
-	outString = strings.Trim(outString[:len(outString)-1], " ")
-	// Remove duplicates to avoid having them in the certificates.
-	names := strings.Split(outString, " ")
-	set := sets.NewString(names...)
-	allHostnames = set.List()
-	return allHostnames, nil
+	return c
 }
 
-func NewMicroshiftConfig() *MicroshiftConfig {
-	nodeName, err := os.Hostname()
+// fillDefaults forcibly sets the configuration to the default
+// values. We do not use a static struct for the defaults because some
+// of them are computed from the environment. If any error occurs
+// probing the environment, the values in the Config instance are not
+// changed.
+func (c *Config) fillDefaults() error {
```

Comment:
Should we use constants for the fixed values in this function?

Reply:
We could. The main point of having constants is to reuse the same value in multiple places. In this case most of these values only appear the one time, except for tests. I could prepare another PR to see what the changes would look like.

Reply:
That was my point, yes. In order to reduce cognitive load when changing tests (or even md docs in the repo), have them all in a single place easy to look for. I am ok with doing it in another PR.